File: 프로그래머스/lv2/81302. 거리두기 확인하기/거리두기 확인하기.py
# P = 응시자
# 0 = 빈 테이블
# X = 파티션
# 거리 2이내에 응시자가 존재하지 않도록

# return 1 or 0

dx = [0, 1, 0, -1]
dy = [1, 0, -1, 0]

def outOfRange(x, y):
    if x < 0 or y < 0 or x >= 5 or y >= 5:
        return True
    return False

# 범위내에 사람이 존재하면 True not False
def isExist(place, x, y, direction, flag):
    for i in range(4):
        if abs(direction - i) == 2: continue
        nx = x + dx[i]
        ny = y + dy[i]
        
        if outOfRange(nx, ny): continue
        
        if place[nx][ny] == 'P' and flag == 'O':
            return True
    return False

def isValid(place):
    for x in range(5):
        for y in range(5):
            if place[x][y] != 'P': continue
            
            for i in range(4):
                nx = x + dx[i]
                ny = y + dy[i]
                
                if outOfRange(nx, ny): continue
                
                if place[nx][ny] == 'P':
                    return 0
                
                if isExist(place, nx, ny, i, place[nx][ny]):
                    return 0
                
    return 1
    
def solution(places):
    answer = []
    for place in places:
        answer.append(isValid(place))
    return answer

File: 프로그래머스/lv2/81302. 거리두기 확인하기/거리두기 확인하기.java
import java.util.*;

// 맨해튼 거리 ( |x1 - x2| + |y1 - y2| > 2 [단, 파티션이 있다면 가능])
// 응시자(P), 빈 테이블(0), 파티션(X)
// 한 번의 4방 탐색 진행

class Solution {
    final static int[] dx = {0, 1, -1, 0};
    final static int[] dy = {1, 0, 0, -1};
    public int[] solution(String[][] places) {
        int[] answer = new int[5];
        
        for(int i = 0; i < 5; i++){
            char[][] place = new char[5][5];
            for(int j = 0; j < 5; j++){
                place[j] = places[i][j].toCharArray();
            }
            
            if(isValid(place)) answer[i] = 1;
        }
        return answer;
    }
    
    static boolean outOfRange(int x, int y){
        if(x < 0 || y < 0 || x >= 5 || y >= 5) return true;
        return false;
    }
    
    static boolean isExist(char[][] place, int x, int y, int direction, char flag){
        for(int i = 0; i < 4; i++){
            if(direction + i == 3) continue;
            
            int nx = x + dx[i];
            int ny = y + dy[i];
            
            if(outOfRange(nx, ny)) continue;
            if(place[nx][ny] == 'P' && flag == 'O') return true;
        }
        return false;
    }
    
    static boolean isValid(char[][] place){
        for(int x = 0; x < 5; x++){
            for(int y = 0; y < 5; y++){
                if(place[x][y] != 'P') continue;
                
                for(int i = 0; i < 4; i++){
                    int nx = x + dx[i];
                    int ny = y + dy[i];
                    
                    if(outOfRange(nx, ny)) continue;
                    
                    if(place[nx][ny] == 'P') return false;
                    if(isExist(place, nx, ny, i, place[nx][ny])) return false;
                }
            }
        }
        return true;
    }
}

